Add epub support (without audio)

JohnThomson opened the pull request:
This is a squashed commit of a sequence of cherry-picks
from the EpubSpike branch, plus one cleanup commit.

Show epub radio only if experimental commands enabled
Handle properly encoded spaces in image src attributes
Show some information and help above preview
Create UI for publishing epub
Extract class BookTestsBase
Origami image widths percent of marginBox
Some stylesheet tweaks
Rearrange basic text and picture page for bilingual
Embed fonts used in stylesheets
Always include custom styles ...
Include thumbnail.png as cover image
Convert image sizes to percent
Make epub with minimal styles
Change case of HTML elements in CSS
Finish passing validator
Got the basics working
Cherry-pick "Got the basics working"
Adding epub button to PublishView.Designer was abandoned.
Cherry-pick Spike of epub export

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/bloombooks/bloomdesktop/791)
<!-- Reviewable:end -->

hatton commented:
Reviewed 33 of 35 files at r1, 2 of 2 files at r2.
Review status: 33 of 35 files reviewed at latest revision, 25 unresolved discussions, all commit checks successful.

---

<sup>**[src/BloomBrowserUI/epub/epubUnpaginated.less, line 3 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_F1jn0Ab64QSIl7uh)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomBrowserUI/epub/epubUnpaginated.less#L3)):</sup>
What does specification of these latin fonts here actually mean? Does the font specified in Bloom's settings still get used if possible? Could add a comment about this.
<!-- Reviewable comment -K-_F1jn0Ab64QSIl7ui:-2096799773 -->

---

<sup>**[src/BloomBrowserUI/epub/epubUnpaginated.less, line 28 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_F6mhweLyru1TYma4)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomBrowserUI/epub/epubUnpaginated.less#L28)):</sup>
I think ```.licenseAndCopyrightBlock img``` is all the spec needed?
<!-- Reviewable comment -K-_F6mhweLyru1TYma5:64327858 -->

---

<sup>**[src/BloomExe/BloomExe.csproj, line 192 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_FwuTswA0alYKKqRz)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/BloomExe.csproj#L192)):</sup>
Does this fit better in Publish?
<!-- Reviewable comment -K-_FwuTswA0alYKKqS-:1496820682 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 44 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_J7pmBQh9ZWg4KVlI)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L44)):</sup>
Yeah, seems unhelpful, having our own copy of this. _book.Storage would look right to me. Or a property that returns it.
<!-- Reviewable comment -K-_J7pmBQh9ZWg4KVlJ:718521999 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 84 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_Ir2BhSGtJz3bClnq)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L84)):</sup>
How is this related to "Fixed Layout"? If that's what it is, that would make a fine name (thought you'd have to switch the polarity).
<!-- Reviewable comment -K-_Ir2BhSGtJz3bClnr:768918503 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 92 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_JVlkSwxeRh6W14SY)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L92)):</sup>
Could you add a comment explaining when this version is used? In an IDE, I could find out myself :-)
<!-- Reviewable comment -K-_JVlkSwxeRh6W14SZ:1811700202 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 100 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_Jo700xm3SbaK4v-U)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L100)):</sup>
Did you consider breaking staging out into its own class? It doesn't seem to belong here.
<!-- Reviewable comment -K-_Jo700xm3SbaK4v-V:-787260716 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 123 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_K0Z9Z33-6a_Pgd2p)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L123)):</sup>
A good place to factor out a method?
<!-- Reviewable comment -K-_K0Z9Z33-6a_Pgd2q:1815371369 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 152 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_LRJlJhwT7whLuMlU)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L152)):</sup>
New method?
<!-- Reviewable comment -K-_LRJlJhwT7whLuMlV:-1734697287 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 154 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_KDofh3dMqnza5GHB)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L154)):</sup>
This section will need to be updated to use the new HtmlDom methods GetIsImgOrSomethingWithBackgroundImage(), GetImageElementUrl() after PR  https://github.com/BloomBooks/BloomDesktop/pull/792
<!-- Reviewable comment -K-_KDofh3dMqnza5GHC:2080972844 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 225 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_LH_LgcE096KsjXnW)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L225)):</sup>
Wants its own method
<!-- Reviewable comment -K-_LH_LgcE096KsjXnX:-1543805131 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 305 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_Ls_15hDDKN6MVOEZ)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L305)):</sup>
This is irrelevant, now, I think, since the conversion to all custom pages?
<!-- Reviewable comment -K-_Ls_15hDDKN6MVOE_:-648786237 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 337 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_MBRJOrL_q35x3xrR)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L337)):</sup>
StackOverflow says that an xpath could do this: ```(//*[contains(concat(" ", normalize-space(@class), " "), " foo ")])```. But whatever you end up with, please put this and similar utility functions as statics in ```HtmlDom``` where we seem to be collecting such things.
<!-- Reviewable comment -K-_MBRJOrL_q35x3xrS:-2126446081 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 391 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_NFuKl9K8_C_sav50)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L391)):</sup>
I fear we'll lose track of this significant (and time-consuming issue) . Can you make sure we have it listed somewhere as needed before production?
<!-- Reviewable comment -K-_NFuKl9K8_C_sav51:583332846 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 397 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_Nh5XWjkU6DI9DVTo)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L397)):</sup>
I think font-family doesn't require quotes?
<!-- Reviewable comment -K-_Nh5XWjkU6DI9DVTp:409860583 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 412 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_O31EXJ_KROmRffRd)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L412)):</sup>
Wants its own file, perhaps in a Bloom.Publish.Epub namespace
<!-- Reviewable comment -K-_O31EXJ_KROmRffRe:1686005772 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 462 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_OxJBZiTyLjVF5klK)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L462)):</sup>
Wants to go to the start, or maybe better, become part of a new class that has the methods that act on it.
<!-- Reviewable comment -K-_OxJBZiTyLjVF5klL:1270640423 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 478 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_ORiYCeRwlLjqA7o5)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L478)):</sup>
This starts with "Get", but is... maybe... more like a "gather" as it modifies _fontNameToFiles.  Feels like it wants its own class, as it is a nice unique thing that isn't tied to epubs.
<!-- Reviewable comment -K-_ORiYCeRwlLjqA7o6:1736499971 -->

---

<sup>**[src/BloomExe/Publish/PublishModel.cs, line 530 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_bK1EqRVlNRP93rDU)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Publish/PublishModel.cs#L530)):</sup>
Shall we put in the analytics now? We'll forget later.
<!-- Reviewable comment -K-_bK1EqRVlNRP93rDV:1068819857 -->

---

<sup>**[src/BloomExe/Publish/PublishView.cs, line 136 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_bc_dkX04QBI9m4jU)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Publish/PublishView.cs#L136)):</sup>
Might as well make that name parallel to the others
<!-- Reviewable comment -K-_bc_dkX04QBI9m4jV:859803226 -->

---

<sup>**[src/BloomExe/Publish/PublishView.cs, line 341 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_bylV1rc_aqL-tViA)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Publish/PublishView.cs#L341)):</sup>
Can we get rid of _originalSaveText? Would be less weird if this was, say, an extension modifying the "original". But here you've just introduced a new stage in the state machine, Epub, so it should have equal standing. The label of the button, if it should really be the same button, wants to just get its name based on the current DisplayMode (or have its label pushed in when the mode changes).
<!-- Reviewable comment -K-_bylV1rc_aqL-tViB:-1629048711 -->

---

<sup>**[src/BloomExe/Publish/PublishView.cs, line 459 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_cvbadmJrVR5hfHJ5)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Publish/PublishView.cs#L459)):</sup>
Should we fix this need to copy in Readium now, or as soon as this gets merged? Please make a card in 3.5 for this if you don't want to do it now. Sounds like you haven't yet tried file://. Could you also try http://localhost.... and let our server supply it?
<!-- Reviewable comment -K-_cvbadmJrVR5hfHJ6:-434251569 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791)
<!-- Sent from Reviewable.io -->

JohnThomson commented:
Review status: 33 of 35 files reviewed at latest revision, 22 unresolved discussions, all commit checks successful.

---

<sup>**[src/BloomBrowserUI/epub/epubUnpaginated.less, line 3 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_F1jn0Ab64QSIl7uh)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomBrowserUI/epub/epubUnpaginated.less#L3)):</sup>
Depends somewhat on the reader. Comment added.
<!-- Reviewable comment -K-dpPjvgQz10xtPD1uO:1476883274 -->

---

<sup>**[src/BloomBrowserUI/epub/epubUnpaginated.less, line 28 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_F6mhweLyru1TYma4)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomBrowserUI/epub/epubUnpaginated.less#L28)):</sup>
I think I just copied this rule from Factory-XMatter.css (278). I'm afraid I just don't remember how I decided which rules were safe/desirable to include.
<!-- Reviewable comment -K-dpwCM77Tn0Bn08xsw:-1097900176 -->

---

<sup>**[src/BloomExe/BloomExe.csproj, line 192 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_FwuTswA0alYKKqRz)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/BloomExe.csproj#L192)):</sup>
Done
<!-- Reviewable comment -K-dqYV6-RZdAPzeYwoN:676635175 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 44 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_J7pmBQh9ZWg4KVlI)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L44)):</sup>
Made a Storage property of EpubMaker returning _book.Storage
<!-- Reviewable comment -K-dr8INXSJS3KjqwhGt:309892907 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 84 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_Ir2BhSGtJz3bClnq)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L84)):</sup>
Well, I'm not sure a fixed layout is really possible in an epub. It sort of contrasts with my initial unsuccessful attempts to make an epub that follows the page layout of the paginated books. If we succeed in coming up with a layout that is closer to the original pagination, and there is some reason not to use it always, the right name for the controlling setting may be clearer.
What would you name the stylesheet that is used for "Not Fixed Layout" if you give the property such a name? UnFixedLayout.css? SimpleLayout.css?
Maybe we should drop the property and the unused behavior when unpaginated is false, and just call the stylesheet defaultEpub.css?
<!-- Reviewable comment -K-drNAzsOR8SX1SILhZ:99432439 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 92 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_JVlkSwxeRh6W14SY)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L92)):</sup>
There's really no good reason to have a constructor at all. I removed both.
<!-- Reviewable comment -K-dtOXv0mI4Jzqs7I-L:1608740096 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 100 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_Jo700xm3SbaK4v-U)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L100)):</sup>
I'm not sure why not. Maybe staging is a bad name. Staging is nearly all the work of making the epub, as reflected in the class name: it assembles all the files that need to be zipped into the epub (and displayed by the preview). All that is left to do to make the epub is to zip the results into the final file.
<!-- Reviewable comment -K-dtX3HmJtVxB_cYc8M:-1541173211 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 123 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_K0Z9Z33-6a_Pgd2p)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L123)):</sup>
Done.
<!-- Reviewable comment -K-dxNoPBV0p5cNa4g3g:-499146009 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 152 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_LRJlJhwT7whLuMlU)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L152)):</sup>
Done.
<!-- Reviewable comment -K-dxmUAY__CdzjmWEIh:-499146009 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 154 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_KDofh3dMqnza5GHB)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L154)):</sup>
Yep
<!-- Reviewable comment -K-dxpDhcygtGxU3u9MT:160394143 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 225 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_LH_LgcE096KsjXnW)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L225)):</sup>
Done.
<!-- Reviewable comment -K-dygMRc0z0creEj2Uw:-499146009 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 305 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_Ls_15hDDKN6MVOEZ)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L305)):</sup>
I guess so...though we might be glad of it if someone downloads an old book and makes an epub without editing it first, so the conversion doesn't get a chance to happen?
<!-- Reviewable comment -K-dyk-kcCG8kfF7Nxp_:-1817621609 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 337 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_MBRJOrL_q35x3xrR)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L337)):</sup>
I moved it. Trying to do something this simple with a somewhat complex xpath doesn't feel worth it to me, but I can try if you want.
<!-- Reviewable comment -K-e-skypl62bNax5v4N:-1038409192 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 391 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_NFuKl9K8_C_sav50)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L391)):</sup>
Added to the list in our document. I'm not sure it is necessary before production, though. At worst, we include some unnecessary fonts. I wouldn't hold up an alpha over that, and I'm not sure I'd hold up a release. If there's a real demand for making epubs with Bloom, and if people care enough about file size, we can optimize the next release...
<!-- Reviewable comment -K-e0c_lyGGtnbwkFK4W:-1600250497 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 397 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_Nh5XWjkU6DI9DVTo)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L397)):</sup>
Only if the font name has spaces in it...which makes for a really nasty regex, since fonts with and without quotes can be mixed. What's worse, this regex will only find the first one in any list. I'll work on it.
<!-- Reviewable comment -K-e1VghSZwCYrCUrV3P:867881026 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 412 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_O31EXJ_KROmRffRd)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L412)):</sup>
Done. (Well, I didn't create a new namespace yet.)
<!-- Reviewable comment -K-e6EYV_Wza424nNk_E:1898203287 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 462 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_OxJBZiTyLjVF5klK)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L462)):</sup>
Made a class FontFileFinder
<!-- Reviewable comment -K-e85h3dMAd9ea7N7z0:163062718 -->

---

<sup>**[src/BloomExe/Book/EpubMaker.cs, line 478 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_ORiYCeRwlLjqA7o5)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Book/EpubMaker.cs#L478)):</sup>
I made the class, but I think the name is good. As far as the client is concerned, what it does is get the files for a particular font. It's an implementation secret that the first call builds a cache.
<!-- Reviewable comment -K-e8CjqRvMEuLDL4JYv:700789695 -->

---

<sup>**[src/BloomExe/Publish/PublishModel.cs, line 530 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_bK1EqRVlNRP93rDU)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Publish/PublishModel.cs#L530)):</sup>
Done.
<!-- Reviewable comment -K-eCWTgZN-UsQWqSzpf:-499146009 -->

---

<sup>**[src/BloomExe/Publish/PublishView.cs, line 136 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_bc_dkX04QBI9m4jU)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Publish/PublishView.cs#L136)):</sup>
Done.
<!-- Reviewable comment -K-eD56IDqRovIJc8pvo:-499146009 -->

---

<sup>**[src/BloomExe/Publish/PublishView.cs, line 341 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_bylV1rc_aqL-tViA)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Publish/PublishView.cs#L341)):</sup>
I've removed it. The price is that the designer and the code above both know the text and localization ID for the default case (though what's in the button itself will never be seen).
<!-- Reviewable comment -K-eEDLNvjVsIUQf8-os:-1197530705 -->

---

<sup>**[src/BloomExe/Publish/PublishView.cs, line 459 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_cvbadmJrVR5hfHJ5)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Publish/PublishView.cs#L459)):</sup>
It seems to work well enough that I don't feel a strong urge to try to fix it. Not sure why I didn't try the file: and localhost: options, if I really didn't. Anyway, I made a card.
<!-- Reviewable comment -K-eFG3HkYiuUGYhjpxU:-1478810127 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791)
<!-- Sent from Reviewable.io -->

JohnThomson commented:
So at this point I think I've responded to all the comments and made the code changes I think you wanted (with maybe a couple of exceptions explained in comments). Back to you.

hatton commented:
<sup>**[src/BloomBrowserUI/epub/epubUnpaginated.less, line 3 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_F1jn0Ab64QSIl7uh)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomBrowserUI/epub/epubUnpaginated.less#L3)):</sup>
If the goal is to track basePage, then let's factor out that material to something that this shares, and @import it.
<!-- Reviewable comment -K06nGJApQDKxixCg2eq:1308862306 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791)
<!-- Sent from Reviewable.io -->

hatton commented:
Review status: 28 of 38 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

<sup>**[src/BloomBrowserUI/epub/epubUnpaginated.less, line 28 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_F6mhweLyru1TYma4)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomBrowserUI/epub/epubUnpaginated.less#L28)):</sup>
OK, if you copied from CSS, so you got a fully expanded rule. Since this is a LESS file, you want the original LESS rule instead.
<!-- Reviewable comment -K06nk9-ekxOgjupIg6N:1779497475 -->

---

<sup>**[src/BloomExe/Book/HtmlDom.cs, line 860 \[r3\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K06oMrz6uvmtaP-A_DH)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b243d3e5476ad145f95c064b3df20f8d1194792d/src/BloomExe/Book/HtmlDom.cs#L860)):</sup>
How about "GetAtributeValue"?
<!-- Reviewable comment -K06oMrz6uvmtaP-A_DI:-1676523356 -->

---

<sup>**[src/BloomExe/Publish/EpubMaker.cs, line 84 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_Ir2BhSGtJz3bClnq)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Publish/EpubMaker.cs#L84)):</sup>
> Well, I'm not sure a fixed layout is really possible in an epub

Hmmm... but isn't that exactly what it's called in epub-land? I think it's called "Fixed Layout". So to make the code track standard epub terminology, would it be correct for this to be "NotFixedLayout"? Or is this just a similar but not actually matching concept?
<!-- Reviewable comment -K06ob8D7kQHe6H81dyk:-1738837186 -->

---

<sup>**[src/BloomExe/Publish/EpubMaker.cs, line 305 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_Ls_15hDDKN6MVOEZ)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Publish/EpubMaker.cs#L305)):</sup>
If you can get this far without triggering all our migration code, then that's probably a number of bugs waiting to happen, right?
<!-- Reviewable comment -K06pcQ1JnDX_phHy-BQ:-1207431236 -->

---

<sup>**[src/BloomExe/Publish/EpubMaker.cs, line 407 \[r3\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K06qlh-SbrjJk3-J_96)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b243d3e5476ad145f95c064b3df20f8d1194792d/src/BloomExe/Publish/EpubMaker.cs#L407)):</sup>
For this new FindFontsUsedInCss() method... I think it would be cheap to go out to HTML dom and get a good unit test?
<!-- Reviewable comment -K06qlh-SbrjJk3-J_97:2071287878 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791)
<!-- Sent from Reviewable.io -->

JohnThomson commented:
Review status: 28 of 38 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

<sup>**[src/BloomBrowserUI/epub/epubUnpaginated.less, line 3 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_F1jn0Ab64QSIl7uh)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomBrowserUI/epub/epubUnpaginated.less#L3)):</sup>
Working on extracting the common bits of basePage.css and Factory-XMatter.css (and trying to remember why, where I have something that isn't in either). I'm afraid this first check-in was just what I came up with that seemed to work OK, rather than a carefully minimal stylesheet.
<!-- Reviewable comment -K07YbEGqg-97vbh5HKH:-1857103702 -->

---

<sup>**[src/BloomExe/Book/HtmlDom.cs, line 860 \[r3\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K06oMrz6uvmtaP-A_DH)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b243d3e5476ad145f95c064b3df20f8d1194792d/src/BloomExe/Book/HtmlDom.cs#L860)):</sup>
Done.
<!-- Reviewable comment -K07UEVO3mDJBxdursUo:-499146009 -->

---

<sup>**[src/BloomExe/Publish/EpubMaker.cs, line 84 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_Ir2BhSGtJz3bClnq)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Publish/EpubMaker.cs#L84)):</sup>
You're right...Fixed Layout is a definite epub concept. But this stylesheet is not a set of unique things that make a normally fixed-layout document not-fixed-layout. It's just a minimal set of common elements which make the book look something like the printed one. It's almost incidental that it's associated with not forcing a fixed layout. If we ever figure out how to make a usefully fixed-layout epub, it will probably use this (or a subset of it) and then something else.

So I'm leaning all the more towards defaultEpub.css or commonEpub.css.
<!-- Reviewable comment -K07UiupA02WqECyrrse:-92822271 -->

---

<sup>**[src/BloomExe/Publish/EpubMaker.cs, line 305 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_Ls_15hDDKN6MVOEZ)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Publish/EpubMaker.cs#L305)):</sup>
I need to look into this more. My recollection is that the migration to origami is distinct from the other migrations in happening only when you open the particular page for editing (presumably since that's when you might want to be able to redesign it).

It's not difficult to take out if you want.
<!-- Reviewable comment -K07VSf-pDfbs7SsdzFQ:-1296534386 -->

---

<sup>**[src/BloomExe/Publish/EpubMaker.cs, line 397 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_Nh5XWjkU6DI9DVTo)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Publish/EpubMaker.cs#L397)):</sup>
Done.
<!-- Reviewable comment -K07YKmNEgP0masLqo1S:-499146009 -->

---

<sup>**[src/BloomExe/Publish/EpubMaker.cs, line 407 \[r3\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K06qlh-SbrjJk3-J_96)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b243d3e5476ad145f95c064b3df20f8d1194792d/src/BloomExe/Publish/EpubMaker.cs#L407)):</sup>
I have unit tests for the current version, but as for switching over to making a DOM...to really do it right we need a DOM that is the epub for the whole book, which we aren't currently building anywhere. Or else move the logic for gathering fonts out of this method into something we do for each page. Either way we have to add JavaScript to the page and probably pull in Jquery and stick the page into a browser so we can run the JS and navigate to it and wait for the navigation to finish and set up an event handler to get the result back from JS land...I'm guessing 1-3 days, which probably means more like a week. So no, I wouldn't call it cheap.

Maybe you have something else in mind which is better than what I have but still cheap? If so I'm not seeing what you mean.
<!-- Reviewable comment -K07Wm3GiX9ObtoAKSRS:904429429 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791)
<!-- Sent from Reviewable.io -->

hatton commented:
Reviewed 11 of 11 files at r3.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

<sup>**[src/BloomExe/Book/HtmlDom.cs, line 860 \[r3\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K06oMrz6uvmtaP-A_DH)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b243d3e5476ad145f95c064b3df20f8d1194792d/src/BloomExe/Book/HtmlDom.cs#L860)):</sup>
Ok... that change isn't showing in reviewable yet.
<!-- Reviewable comment -K08HU3lFwq69rYN-H6T:-557766099 -->

---

<sup>**[src/BloomExe/Publish/EpubMaker.cs, line 84 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_Ir2BhSGtJz3bClnq)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Publish/EpubMaker.cs#L84)):</sup>
OK, thanks for explaining.
<!-- Reviewable comment -K08HdSt6P1n9RUrbCbB:349224682 -->

---

<sup>**[src/BloomExe/Publish/EpubMaker.cs, line 305 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_Ls_15hDDKN6MVOEZ)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Publish/EpubMaker.cs#L305)):</sup>
Yes, I think it makes more sense to do that migration before we get to this point. Where we have had these "change only when the user brings up the page in edit" fixes, there has been understandable confusion for the Uganda folks when they go straight to PDF, or to the librarian when she goes straight to upload.
<!-- Reviewable comment -K08Hu5LJE89V_7Pg23L:589387177 -->

---

<sup>**[src/BloomExe/Publish/EpubMaker.cs, line 407 \[r3\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K06qlh-SbrjJk3-J_96)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b243d3e5476ad145f95c064b3df20f8d1194792d/src/BloomExe/Publish/EpubMaker.cs#L407)):</sup>
Sorry, tripped you up there by writing "Html dom" instead of "HtmlDom", as in the class where we're collecting various html-related methods. YOu could argue that this is css, not HTML, which is true, except it shows u in htmls files. Hmmm.  I just like to see generic things land in generic places :-)
<!-- Reviewable comment -K08IgtSk69IGAi2xsXb:-2112624243 -->

---

<sup>**[src/BloomExe/Publish/EpubMaker.cs, line 478 \[r3\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K08KS_awiLqsC-_DITk)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b243d3e5476ad145f95c064b3df20f8d1194792d/src/BloomExe/Publish/EpubMaker.cs#L478)):</sup>
This is still showing as "AttrVal" instead of "GetAttributeValue" in reviewable.
<!-- Reviewable comment -K08KS_awiLqsC-_DITl:-1425942491 -->

---

<sup>**[src/BloomExe/Publish/EpubMaker.cs, line 661 \[r3\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K08Lgk-D0lplNWYzqA1)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b243d3e5476ad145f95c064b3df20f8d1194792d/src/BloomExe/Publish/EpubMaker.cs#L661)):</sup>
Should this be UrlPathEncode() instead? I think (not 100% sure) that UrlEncode converts a space to '+' instead of %20.
<!-- Reviewable comment -K08Lgk-D0lplNWYzqA2:-938514548 -->

---

<sup>**[src/BloomExe/Publish/PublishView.cs, line 341 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_bylV1rc_aqL-tViA)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Publish/PublishView.cs#L341)):</sup>
Suppose you'd want to remove that from the designer then?
<!-- Reviewable comment -K08MMtVmenHA4F_aGN3:1316301686 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791)
<!-- Sent from Reviewable.io -->

JohnThomson commented:
Review status: 27 of 49 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

<sup>**[src/BloomExe/Book/HtmlDom.cs, line 860 \[r3\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K06oMrz6uvmtaP-A_DH)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b243d3e5476ad145f95c064b3df20f8d1194792d/src/BloomExe/Book/HtmlDom.cs#L860)):</sup>
Should be when I publish this next round of comments.
<!-- Reviewable comment -K0CA6g4AJ1B7c4ajetz:-494501369 -->

---

<sup>**[src/BloomExe/Publish/EpubMaker.cs, line 84 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_Ir2BhSGtJz3bClnq)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Publish/EpubMaker.cs#L84)):</sup>
I have renamed epubUnpaginated to baseEpub
<!-- Reviewable comment -K0CACIAKAzdMMVGy-va:1003805524 -->

---

<sup>**[src/BloomExe/Publish/EpubMaker.cs, line 305 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_Ls_15hDDKN6MVOEZ)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Publish/EpubMaker.cs#L305)):</sup>
I've removed the special case. I think re-arranging the way we migrate books and pages should be a separate task, PR, and prioritization.
<!-- Reviewable comment -K0CLc0oEYhYzdFI4pMx:885583125 -->

---

<sup>**[src/BloomExe/Publish/EpubMaker.cs, line 407 \[r3\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K06qlh-SbrjJk3-J_96)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b243d3e5476ad145f95c064b3df20f8d1194792d/src/BloomExe/Publish/EpubMaker.cs#L407)):</sup>
Done.
<!-- Reviewable comment -K0CEwYnPAombKPPAne1:-499146009 -->

---

<sup>**[src/BloomExe/Publish/EpubMaker.cs, line 478 \[r3\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K08KS_awiLqsC-_DITk)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b243d3e5476ad145f95c064b3df20f8d1194792d/src/BloomExe/Publish/EpubMaker.cs#L478)):</sup>
Will finish commit process before next publish.  Sorry.
<!-- Reviewable comment -K0CEy2UHR1ku0QW9Mf8:-339729198 -->

---

<sup>**[src/BloomExe/Publish/EpubMaker.cs, line 661 \[r3\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K08Lgk-D0lplNWYzqA1)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b243d3e5476ad145f95c064b3df20f8d1194792d/src/BloomExe/Publish/EpubMaker.cs#L661)):</sup>
You are right about the plus. There is no WebUitlity.UrlPathEncode. There is a HttpUtility.UrlPathEncode but it is deprecated without explanation. Uri.EscapeDataString seems to do the right thing??
<!-- Reviewable comment -K0CIM32CkZMUq1EFPTo:-1583612132 -->

---

<sup>**[src/BloomExe/Publish/PublishView.cs, line 341 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_bylV1rc_aqL-tViA)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Publish/PublishView.cs#L341)):</sup>
I doubt the designer will allow the button not to have any text. I changed it to "Save stub" as some indication that this is not expected to be seen, and removed the designer code for localizing it, though I wouldn't be surprised if some version comes back the next time someone runs the designer on this control.
<!-- Reviewable comment -K0CJeqx57fwWMbwh0H0:-756300902 -->

---

<sup>**[src/BloomBrowserUI/epub/epubUnpaginated.less, line 28 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_F6mhweLyru1TYma4)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomBrowserUI/epub/epubUnpaginated.less#L28)):</sup>
I have extracted the common stuff in basePage.less to a new "CommonBasePageEpub".less and the common stuff in bloom-xmatter-common.less to BloomXMatterEpubCommon.less. I think I'm now using the preferred .less coding style.
<!-- Reviewable comment -K0C9XJGkybtbpScTMgD:-1385200987 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791)
<!-- Sent from Reviewable.io -->

hatton commented:
Reviewed 22 of 24 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[src/BloomExe/Book/HtmlDom.cs, line 868 \[r4\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K0CPPdd0fwafCS73_-k)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/1e586644ba2b3b77a2fbc2cfe007a1a511295408/src/BloomExe/Book/HtmlDom.cs#L868)):</sup>
What's the value of "internal" in something like Bloom, which is not a library?
<!-- Reviewable comment -K0CPPdeNZrWT6QWjiBt:-1043383983 -->

---

<sup>**[src/BloomExe/Publish/EpubMaker.cs, line 661 \[r3\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K08Lgk-D0lplNWYzqA1)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b243d3e5476ad145f95c064b3df20f8d1194792d/src/BloomExe/Publish/EpubMaker.cs#L661)):</sup>
I'd stick with HttpUtility.UrlPathEncode for now
<!-- Reviewable comment -K0CV0sD8kpcqXv8GtOu:1396829652 -->

---

<sup>**[src/BloomExe/Publish/PublishView.cs, line 341 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_bylV1rc_aqL-tViA)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Publish/PublishView.cs#L341)):</sup>
I was thinking of removing the l10n id, that's all. Not touching code or the label. Nothing more needed now, though.
<!-- Reviewable comment -K0CPnIn8ujoLgW4odOg:-879983127 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791)
<!-- Sent from Reviewable.io -->

JohnThomson commented:
Review status: 47 of 49 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

<sup>**[src/BloomExe/Book/HtmlDom.cs, line 868 \[r4\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K0CPPdd0fwafCS73_-k)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/1e586644ba2b3b77a2fbc2cfe007a1a511295408/src/BloomExe/Book/HtmlDom.cs#L868)):</sup>
I sometimes use it for "I would like to make this private except that I want to directly write unit tests for it" though I might also use it for "this is a special-purpose method that is intended for use only by selected 'friend' classes". This was originally a case of the former but as we are promoting it to HtmlDom I think it should also be promoted to public. I will do that.
<!-- Reviewable comment -K0C_R5yv0aAFvN7t5qv:-31276835 -->

---

<sup>**[src/BloomExe/Publish/EpubMaker.cs, line 661 \[r3\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K08Lgk-D0lplNWYzqA1)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/b243d3e5476ad145f95c064b3df20f8d1194792d/src/BloomExe/Publish/EpubMaker.cs#L661)):</sup>
Done.
<!-- Reviewable comment -K0Ca9m5H4x7c9mizKhf:-499146009 -->

---

<sup>**[src/BloomExe/Publish/PublishView.cs, line 341 \[r2\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791#-K-_bylV1rc_aqL-tViA)** ([raw file](https://github.com/bloombooks/bloomdesktop/blob/517b5eef15e9f06b2051185e89ed67f7634e3beb/src/BloomExe/Publish/PublishView.cs#L341)):</sup>
OK
<!-- Reviewable comment -K0C_GfVaRfUqSDQL6nz:5173665 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/791)
<!-- Sent from Reviewable.io -->
